File: practice/scala/easy/0121_BestTimeToBuyAndSellStock.scala
object Solution {

  def maxProfit(prices: Array[Int]): Int = {
    // return type of .foldLeft will be (minCost, maxProfit)
    prices.foldLeft((Int.MaxValue, 0)) {
      // return type of tuple with the current price
      case ((minMin, maxProfit), price) =>
        val curMin = if (price < minMin) price else minMin
        val curProfit = 
            if (price - curMin > maxProfit) price - curMin
            else maxProfit
        (curMin, curProfit)
    }._2
  }
}

File: practice/easy/0121-Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
  def maxProfit(self, prices: List[int]) -> int:
    if len(prices) == 0:
        return 0

    cur_min = prices[0]
    cur_max = prices[0]
    global_max = 0

    for price in prices[1:]:
      # if this price is lower than the lowest price we've seen so far,
      # restart the search for a local max here
      if price < cur_min:
        cur_min = price
        cur_max = price
      # if this price is greater than the local max, update the local max
      elif price > cur_max:
        cur_max = price

      global_max = max(global_max, cur_max-cur_min)

    return global_max
